cgen not creating generic interface casting function when passing type param to call expr

<!-- Please make sure to run `v up` before reporting any issues as it may have already been fixed.
     It's also advisable to update all relevant modules using `v outdated` and `v install` -->

<!-- You can use `v doctor` to fill up the next fields -->
```
OS: macos, macOS, 11.6, 20G165
Processor: 8 cpus, 64bit, little endian, Apple M1
CC version: Apple clang version 13.0.0 (clang-1300.0.29.3)

getwd: /Users/nedpals/Documents/coding/vls
vmodules: /Users/nedpals/.vmodules
vroot: /Users/nedpals/Downloads/v
vexe: /Users/nedpals/Downloads/v/v
vexe mtime: 2022-06-06 10:40:11
is vroot writable: true
is vmodules writable: true
V full version: V 0.2.4 8a2236d.7f67981

Git version: git version 2.33.0
Git vroot status: weekly.2022.22-57-g7f679816 (2 commit(s) behind V master)
.git/config present: true
thirdparty/tcc status: thirdparty-macos-arm64 173c526e
```

<!-- Please include the standalone minimal reproducible code.
If not, it may be closed by the administrators. -->
**What did you do?**
This example is an extension of #14692.

Supposed C function for casting generic interface type into it's desired type is not automatically created when the type parameter `T` is inherited from the function declaration to the call expressions. 

```v
module main

interface TypeFactory<T> {
   get_type(type_name string) T
}

enum NodeType {
   unknown
   expression
   statement
   literal
}

struct EnumTypeFactory {}
fn (f &EnumTypeFactory) get_type(type_name string) NodeType {
   return match type_name {
      'expression' { NodeType.expression }
      'statement' { NodeType.statement }
      'literal' { NodeType.literal }
      else { NodeType.unknown }
   }
}

struct RawNode {
   type_name string
}

struct Node<T> {
   factory TypeFactory<T>
   type_name NodeType
   raw_node  RawNode
}

fn new_node<T>(factory TypeFactory<T>, raw_node RawNode) ?Node<T> {
   return Node<T>{
     factory: factory
     type_name: factory.get_type(raw_node.type_name)
     raw_node: raw_node
   }
}

fn program<T>(factory TypeFactory<T>) ? {
   root := new_node<T>(factory, RawNode{'literal'})?
   println(root)   
   println(new_node<T>(root.factory, RawNode{'expression'})?)
}

fn main() {
   program<NodeType>(&EnumTypeFactory{})?
}
```

**What did you expect to see?**
```
Node<NodeType>{
    factory: TypeFactory<NodeType>(EnumTypeFactory{})
    type_name: literal
}
Node<NodeType>{
    factory: TypeFactory<NodeType>(EnumTypeFactory{})
    type_name: expression
}
```
 
**What did you see instead?**
cgen error:
```
==================
   ^~~~~~~~~~~~~~~
/tmp/v_501/test.7390106379377841584.tmp.c:11686:97: error: implicit declaration of function 'I_main__TypeFactory_T_main__NodeType_to_Interface_main__TypeFactory_T_main__NodeType' is invalid in C99 [-Werror,-Wimplicit-function-declaration]
        _option_main__Node_T_main__NodeType _t1 = main__new_node_T_main__NodeType(/*&main.TypeFactory*/I_main__TypeFactory_T_main__NodeType_to_Interface_main__TypeFactory_T_main__NodeType(&factory), ((main__RawNode){.type_name = _SLIT("literal"),}));
                                                                                                       ^
/tmp/v_501/test.7390106379377841584.tmp.c:11686:97: note: did you mean 'I_main__EnumTypeFactory_to_Interface_main__TypeFactory_T_main__NodeType'?
/tmp/v_501/test.7390106379377841584.tmp.c:2032:50: note: 'I_main__EnumTypeFactory_to_Interface_main__TypeFactory_T_main__NodeType' declared here
static inline main__TypeFactory_T_main__NodeType I_main__EnumTypeFactory_to_Interface_main__TypeFactory_T_main__NodeType(main__EnumTypeFactory* x) {
                                                 ^
/tmp/v_501/test.7390106379377841584.tmp.c:11686:97: error: passing 'int' to parameter of incompatible type 'main__TypeFactory_T_main__NodeType' (aka 'struct main__TypeFactory_T_main__NodeType')
        _option_main__Node_T_main__NodeType _t1 = main__new_node_T_main__NodeType(/*&main.TypeFactory*/I_main__TypeFactory_T_main__NodeType_to_Interface_main__TypeFactory_T_main__NodeType(&factory), ((main__RawNode){.type_name = _SLIT("literal"),}));
                                                                                                       ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/tmp/v_501/test.7390106379377841584.tmp.c:11679:120: note: passing argument to parameter 'factory' here
...
==================
```

--- a/vlib/v/gen/c/cgen.v
+++ b/vlib/v/gen/c/cgen.v
@@ -2126,8 +2126,8 @@ fn (mut g Gen) expr_with_cast(expr ast.Expr, got_type_raw ast.Type, expected_typ
 		g.expr(expr)
 		return
 	}
-	if exp_sym.info is ast.Interface && got_type.idx() != expected_type.idx()
-		&& !expected_type.has_flag(.optional) {
+	if got_sym.info !is ast.Interface && exp_sym.info is ast.Interface
+		&& got_type.idx() != expected_type.idx() && !expected_type.has_flag(.optional) {
 		if expr is ast.StructInit && !got_type.is_ptr() {
 			g.inside_cast_in_heap++
 			got_styp := g.cc_type(got_type.ref(), true)

--- a/vlib/v/tests/generics_interface_with_non_generic_method_test.v
+++ b/vlib/v/tests/generics_interface_with_non_generic_method_test.v
@@ -13,7 +13,7 @@ enum NodeType {
 
 struct EnumTypeFactory {}
 
-fn (f EnumTypeFactory) get_type(type_name string) NodeType {
+fn (f &EnumTypeFactory) get_type(type_name string) NodeType {
 	return match type_name {
 		'expression' { NodeType.expression }
 		'statement' { NodeType.statement }
@@ -22,24 +22,34 @@ fn (f EnumTypeFactory) get_type(type_name string) NodeType {
 	}
 }
 
+struct RawNode {
+	type_name string
+}
+
 struct Node<T> {
 	factory   TypeFactory<T>
 	type_name NodeType
+	raw_node  RawNode
 }
 
-fn new_node<T>(type_name string, factory TypeFactory<T>) Node<T> {
+fn new_node<T>(factory TypeFactory<T>, raw_node RawNode) ?Node<T> {
 	return Node<T>{
 		factory: factory
-		type_name: factory.get_type(type_name)
+		type_name: factory.get_type(raw_node.type_name)
+		raw_node: raw_node
 	}
 }
 
-fn test_generic_interface_with_non_generic_method() {
-	root1 := new_node<NodeType>('literal', EnumTypeFactory{})
+fn program<T>(factory TypeFactory<T>) ? {
+	root1 := new_node<T>(factory, RawNode{'literal'})?
 	println(root1)
 	assert root1.type_name == .literal
 
-	root2 := new_node<NodeType>('expression', root1.factory)
+	root2 := new_node<T>(root1.factory, RawNode{'expression'})?
 	println(root2)
 	assert root2.type_name == .expression
+}
+
+fn test_generic_interface_with_non_generic_method() ? {
+	program<NodeType>(&EnumTypeFactory{})?
 }

